src/module.h

```diff
@@ -136,6 +138,18 @@ bool debugCommandsEnabled(RedisModuleCtx *ctx);
 specialCaseCtx *prepareOptionalTopKCase(const char *query_string, RedisModuleString **argv, int argc, uint dialectVersion,
                              QueryError *status);
 
+/**
+ * Get the effective K value for a KNN context based on whether this is a shard command.
+ * For shard commands, applies shard window ratio optimization.
+ * For coordinator commands, returns original K.
+ */
+static inline size_t getEffectiveK(const knnContext *ctx, bool isShardCommand) {
+  if (!isShardCommand || ctx->shardWindowRatio >= 1.0) {
```

Comment:
`ctx->shardWindowRatio >= 1.0` should never happen if we validate earlier in the pipeline?